### PR TITLE
Add gaps with forced ventilation

### PR DESCRIPTION
**Pull request by simonwacker**
Windows-CalcEngine can calculate the speed of the gas in the gap between layers caused by natural convection. When mechanical ventilation is used, the speed of the gas in the gap can be set to a constant value. This pull request adds the functionality of forced ventilation in the gap. In this case, the speed of the gas in the gap is simply set to a constant.

### CODE REVIEW

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    SCOPED_TRACE("Begin Test: Test Forced Ventilated Gap Layer - Airflow Reference Point");
+
+    auto aLayer = GetGap();
+
+    // Airflow iterations are set to 1e-4 and it cannot exceed that precision
+
+    ASSERT_TRUE(aLayer != nullptr);
+    auto airflowReferencePoint = aLayer->getAirflowReferencePoint(0.5);
+    EXPECT_NEAR(6911.7392493523666, airflowReferencePoint, 1e-4);
+}
```
Comment (simonwacker) [inline src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp]: @bbuenoun Do these tests and numbers make sense?

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+protected:
+    void SetUp() override
+    {
+        /////////////////////////////////////////////////////////
+        // Outdoor
+        /////////////////////////////////////////////////////////
+        auto airTemperature = 255.15;   // Kelvins
+        auto airSpeed = 5.5;            // meters per second
+        auto tSky = 255.15;             // Kelvins
+        auto solarRadiation = 700.0;
```
Comment (simonwacker) [inline src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp]: @bbuenoun In this test I set the solar radiation to `700`. The computed numbers though do **not** change compared to [the test with solar radiation `0`](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/files#diff-f1510538eafc05022ed4282de85aba1f6ce25356865aa7c328620229546650a7R20).

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+        EffectiveLayers::EffectiveOpenness effOpenness{effectiveLayer.getEffectiveOpenness()};
+
+        auto shadeLayer = Tarcog::ISO15099::Layers::shading(
+          shadeLayerThickness, shadeLayerConductance, effOpenness);
+
+        ASSERT_TRUE(shadeLayer != nullptr);
+
+        auto gapThickness = 0.0127;
+        auto gapAirSpeed = 0.0;
```
Comment (simonwacker) [inline src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp]: @bbuenoun In this test I set the gap air speed to `0`. Compared to [the test with the value `0.5`](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/files#diff-f1510538eafc05022ed4282de85aba1f6ce25356865aa7c328620229546650a7R67) only the gain energy changes from `-173.19394352133912` to `0.0`. The reason is that the way the computation currently works, we **force** an air speed of `0` no matter what in [`CIGUShadeLayer::calcEdgeShadeFlow`](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/files#diff-373f0b5a3bab54faaf37cfa2a370957a33143cf5af21a40aa39861baea9356e1R297-R300) (any possible natural ventilation is suppressed). And this air speed of `0` is [used as multiplicative factor in the calculation of `CIGUVentilatedGapLayer::ventilatedFlow`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/IGUVentilatedGapLayer.cpp#L182) resulting in the ventilated flow `0`.

**Review by simonwacker — COMMENTED** [review]
I implemented two other tests, [one with non-zero solar radiation](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/2cc2a8cc9105b252f545f38f4dd7803480ec84fe) and [another one with zero air speed](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/42ce33dad40053dd64fa8a2ad4f1750e6dbb4ffb). I also [added additional tests for the solid and shade layer temperatures](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/1f81793390f794c535a38813414707e789c236e0). As you expected, the front temperature of the gap is the back temperature of the solid layer and the back temperature of the gap is the front temperature of the shade layer.

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp (hunk 3 above)
Comment (bbuenoun) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp]: We have to compare the test with air_velocity=0 with a test without natural ventilation, i.e. sealed and narrow gap between two layers.

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp (hunk 2 above)
Comment (bbuenoun) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp]: We have to check the (hidden) optical properties of the layers. The absorptivity might be set to zero.

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp (hunk 1 above)
Comment (bbuenoun) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp]: What is the inlet temperature of the gap: the indoor temperature or the outdoor temperature? in a case with only two layers, both possibilities are possible

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp (hunk 2 above)
Comment (simonwacker) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp]: In https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/070c3b13645b285690d2f9fd627436226039b7fc I set the absorptance to a non-zero value, called `solve` on the Tarcog system in the set-up, and also tested radiosity values. Without calling `solve`, setting the absoptance didn't have an effect on the values. With calling `solve`, all values changed. I don't really know what solve does. In the code it looks like the system has an initial state and calling solve makes the system proceed to the next stable state. Its implementation is https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/NonLinearSolver.cpp#L54

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp (hunk 3 above)
Comment (simonwacker) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp]: In https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/4bda8f0a9a9848b4230f686345c518d9cd07ebdb I added a test with a naturally ventilated [narrow](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/4bda8f0a9a9848b4230f686345c518d9cd07ebdb#diff-77b34fc1601b2388bada4d0314de1afe488763836024690d16cc0a15b08d9f92R66) gap at the edge. The [gain flow](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/4bda8f0a9a9848b4230f686345c518d9cd07ebdb#diff-77b34fc1601b2388bada4d0314de1afe488763836024690d16cc0a15b08d9f92R116) is close to zero.

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp (hunk 1 above)
Comment (simonwacker) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp]: [The environment used to calculate in and out temperatures of a ventilated gap is the one adjacent to the shade layer adjacent to the gap](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/BaseShade.cpp#L168-L180). So, in the layering `outdoor - solid - gap - shade - indoor`, it is the indoor environment and in the layering `outdoor - shade - gap - solid - indoor`, it is the outdoor environment. For those setups, the temperatures are set on the lines https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/BaseShade.cpp#L334-L347 (Looking closer at this, I am unsure if I implemented it correctly for forced ventilation. In this implementation, `m_inTemperature` is set to `TgapOut` or `tempEnvironment` [depending on which of the two is greater](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/BaseShade.cpp#L334) and [depending on whether `forcedVentilation.VerticalDirection` is up or down](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/IGUVentilatedGapLayer.cpp#L75-L82).). Note that [`tempEnvironment`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/BaseShade.cpp#L303) is the gas temperature of the environment adjacent to the shade layer.

Comment (simonwacker) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp]: The new numbers with the promised changes can be seen on the green lines in https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/7dc4422ceebcdeb37ac8a5cdca8220a1255693a1#diff-f1510538eafc05022ed4282de85aba1f6ce25356865aa7c328620229546650a7

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp (hunk 2 above)
Comment (simonwacker) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp]: The new numbers with about absorptance 10% for the glass layer and absorptance 70% for the shade layer can be seen on the green lines in https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/7dc4422ceebcdeb37ac8a5cdca8220a1255693a1#diff-02f7c1e784833ccc1b166b36f337d7a5d836ed05759ab60fa86880260d995ef0

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp (hunk 3 above)
Comment (simonwacker) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp]: I made a naive attempt to implement sealed forced-ventilation gaps in https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/156b5e035530a00945bf5bb65829df06d1fd703b I currently use the previous layer's temperature as inlet temperature for the gap. Should one be able to configure which layer is used for the inlet temperature? Or should it always be the previous or next layer? I added a test with a [normal sealed forced-ventilation gap](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/156b5e035530a00945bf5bb65829df06d1fd703b#diff-2bc71bc9c8ff5c69e24cd6a17dcce54e9424c2df4496d2976ae854b5a6354b4a), one [such that is narrow](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/156b5e035530a00945bf5bb65829df06d1fd703b#diff-f682c31f98199791424b8439fff92ca8201334d507c0f6f96b8d839cc7c143f2), and one [such that is narrow and has air speed zero](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/aca8734711b4c9721d6a919f5a92ea06fbe7dfd0). Which test should produce similar numbers as [GapLayerAtEdgeForcedVentilationWithZeroAirSpeed](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp)? [GapLayerSealedForcedVentilationNarrow](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerSealedForcedVentilationNarrow.unit.cpp)?

File: src/Tarcog/src/IGUVentilatedGapLayer.cpp [hunk 4]
```diff
@@ -147,10 +159,77 @@ namespace Tarcog
...
+            while(!converged)
+            {
+                double TavGap = averageTemperature();
+                // To calculate `beta` we do not invoke `betaCoeff` because
+                // doing so results in an infinite recursion as it calls
+                // `calculateLayerHeatFlow`, which calls
+                // `calculateConvectionOrConductionFlow`, which calls this
+                // method.
+                double beta = exp(-m_Height / characteristicHeight());
+                double alpha = 1 - beta;
```
Comment (simonwacker) [inline src/Tarcog/src/IGUVentilatedGapLayer.cpp]: Do these variables change during the loop? If not, extract them!

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp (hunk 2 above)
Comment (bbuenoun) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp]: Please modify this unit test (let's call it "summer unit test") with the following boundary conditions:
Irradiation: 783 W/m2.
Outdoor temperature: 305.15 K.
Indoor temperature: 297.15 K
Please indicate what are the (hidden) solar reflectance and solar transmittance of the solid and shade layers, respectively.

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp (hunk 3 above)
Comment (bbuenoun) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp]: [GapLayerAtEdgeForcedVentilationWithZeroAirSpeed](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp) has to be compared with a configuration where all the gaps (Atop, Abottom, Afront, Aright, Aleft) are set to zero.

Comment (simonwacker) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp]: As of https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/3e1e6a4c6507a941cef18a03726f90bfee6e8ea9 sealed forced-ventilation gaps are now modelled by putting a closed shade to one side of a forced-ventilation gap. For a code example see https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/3e1e6a4c6507a941cef18a03726f90bfee6e8ea9#diff-2bc71bc9c8ff5c69e24cd6a17dcce54e9424c2df4496d2976ae854b5a6354b4aR44-R46 and https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/3e1e6a4c6507a941cef18a03726f90bfee6e8ea9#diff-2bc71bc9c8ff5c69e24cd6a17dcce54e9424c2df4496d2976ae854b5a6354b4aR57. The gain energies of [GapLayerAtEdgeForcedVentilationWithZeroAirSpeed](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp) and [GapLayerSealedForcedVentilationNarrow](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerSealedForcedVentilationNarrow.unit.cpp) differ only slightly. The temperatures however differ by about 3 to 5 degrees Celsisus. The gain energies of [GapLayerAtEdgeForcedVentilationWithZeroAirSpeed](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp) compared to [GapLayerSealedForcedVentilation](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerSealedForcedVentilation.unit.cpp) differs by about 132. The temperatures also differ by 3 to 5 degrees Celsius.

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp (hunk 2 above)
Comment (simonwacker) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp]: I added the requested test with summer temperatures and irradiation: [GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer.unit.cpp). I also specified transmittance and emissivity values explicitely in https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer.unit.cpp#L40-L44 and https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer.unit.cpp#L57-L61 and noted how these values are used to calculate the reflectance in [`Surface.cpp#calculateReflectance`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Surface.cpp#L115-L126). The implicit defaults used before are defined in [`Layers.hpp#solid`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Layers.hpp#L22-L27) and [`Layers.hpp#shading`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Layers.hpp#L34-L42).

Comment (bbuenoun) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp]: You specified the infrared transmittance (solidLayerFrontIRTransmittance), but we need the SolarTransmittance (the variable might be called "solidLayerFrontSolarTransmittance"), but I'm not sure

Comment (simonwacker) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp]: There are no other arguments to specify the transmittance that I could find. May it be that the argument given to [the helper function `solid`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Layers.cpp#L14-L26) as well as to [the constructor of solid layers](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/IGUSolidLayer.cpp#L41-L55) have the wrong name (I used the names of those arguments to name my variables in the tests)? In [the helper function `shading`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Layers.cpp#L63-L93) the arguments used in the very same way are named `frontTransmittance` and `backTransmittance` (these arguments are in the case that the shade layer is closed even passed to the helper function `solid` as `...IRTransmittance`; so at least here is an inconsistency in the usage). In the case that the shade layer is open, the transmittance arguments are in both usages passed to [the constructor of `CSurface`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Surface.cpp#L28-L37) in which the argumet is named `t_Transmittance`.

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp (hunk 3 above)
Comment (bbuenoun) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp]: Please compare [GapLayerAtEdgeForcedVentilationWithZeroAirSpeed](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp) with a configuration with the same geometry, naturally ventilated, where all the gaps (Atop, Abottom, Afront, Aright, Aleft) are set to zero.

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp (hunk 2 above)
Comment (bbuenoun) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp]: Ok. We need to discuss with Simon Vadanovic how the shortwave radiation problem is solved in Wincalc and what is the meaning of "solidLayerSolarAbsorptance", whether it already takes into account what is transmitted from the first layer.

Comment (bbuenoun) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp]: The results of [GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer.unit.cpp) are plausible

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp (hunk 1 above)
Comment (bbuenoun) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp]: The results of the winter unit tests are plausible

Comment (bbuenoun) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp]: Please indicate the values of these variables in winter and summer unit tests:
m_HInput, m_HCoefficientModel, m_IRCalculatedOutside

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp (hunk 3 above)
Comment (simonwacker) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp]: I [added a test for the requested configuration](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/5c6fdf145db1467262532c1768c4da5f1ddb8faa) (sorry for being so stubborn) and compared the numbers. The temperatures only differ in the third decimal place and the gain energy differs by `0.0182...`.

File: src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp (hunk 1 above)
Comment (simonwacker) [reply src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp]: I added tests for those members in https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/3492a6d8f5966ba296555b618867719a9d401239 To do so, I had to add `get`-functions for those members to `CEnvironment` and to `CSystem`. In both cases, `m_HInput` and `m_IRCalculatedOutside` is `0` and `m_HCoefficientModel` is `BoundaryConditionsCoeffModel::CalculateH`.

**Comment by simonwacker** [comment]
@christoph-maurer I merged Simon's refactorings. Just pull the latest version and then you can start improving and adding further examples in the form of tests.

**Comment by simonwacker** [comment]
@vidanovic I reviewed your refactorings, really like them, and merged them into this feature branch. The changes add clarity, decouple functionalities and put them into the right place. Thanks! I didn't have the insight into the code base to be able to do the implementation like that.

**Comment by vidanovic** [comment]
@simon-wacker It does look good. I have also found very helpful things that you did too. It did help me to perform this refactoring. Few days before the vacation I was trying to make python equivalent of the algorithm, but my results were not matching. I might spend more time in trying to determine what is causing the difference once we are back to work (Tuesday, January 17th 2023).

**Comment by christoph-maurer** [comment]
@vidanovic In order to validate the implementation of the forced convection, we have compared natural convection and forced convection with zero air speed both with a gap thickness of 1 mm in #9. The validation was successful, the results of both tests are very similar.

**Comment by vidanovic** [comment]
@christoph-maurer Great. I am going to take a quick look and merge it. Thanks!!!

**Review by vidanovic — APPROVED** [review]
Results do make sense to me. Unless someone else objects to this, I would merge this into the main and then merge it into the our main LBNL branch.

**Comment by simonwacker** [comment]
> Results do make sense to me. Unless someone else objects to this, I would merge this into the main and then merge it into the our main LBNL branch.

There is the sibling merge request https://github.com/LBNL-ETA/Windows-CalcEngine/pull/144

**Comment by simonwacker** [comment]
The sibling merge request https://github.com/LBNL-ETA/Windows-CalcEngine/pull/144 was merged into the official repository. :tada:
